driver-core/src/main/com/mongodb/client/model/bulk/ClientWriteModel.java

```diff
@@ -0,0 +1,355 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.client.model.bulk;
+
+import com.mongodb.MongoNamespace;
+import com.mongodb.annotations.Sealed;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.Updates;
+import com.mongodb.internal.client.model.bulk.ClientDeleteManyModel;
+import com.mongodb.internal.client.model.bulk.ClientDeleteOneModel;
+import com.mongodb.internal.client.model.bulk.ClientInsertOneModel;
+import com.mongodb.internal.client.model.bulk.ClientReplaceOneModel;
+import com.mongodb.internal.client.model.bulk.ClientUpdateManyModel;
+import com.mongodb.internal.client.model.bulk.ClientUpdateOneModel;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
+/**
+ * An individual write operation to be executed as part of a client-level bulk write operation.
+ *
+ * @since 5.3
+ */
+@Sealed
+public interface ClientWriteModel {
+    /**
+     * Creates a model for inserting the {@code document} into the {@code namespace}.
+     *
+     * @param namespace The namespace.
+     * @param document The document.
+     * @return The requested model.
+     * @param <TDocument> The document type, for example {@link Document}.
+     * @see Filters
+     */
+    static <TDocument> ClientWriteModel insertOne(
```

Comment:
I wonder whether there will be use cases that we uncover where applications want to create models and subsequently examine them by calling getters.  By returning `ClientWriteModel` from all the factory method, we can not add this without breaking binary compatibility.

We could hedge our bets by creating interface types for each model and returning them from these methods.  They don't even have to contain any methods, but since the types are Sealed we could always add them later, e.g.

```
public interface ClientInsertOneModel extends ClientWriteModel {}
public interface ClientUpdateOneModel extends ClientWriteModel {}
...
```

Reply:
> applications want to create models and subsequently examine them by calling getters

I believe that there is unlikely to be a good reason to do so. But we can definitely do what you proposed. I'll hold off to doing the change for now, in case some other concerns result in a more drastic changes.

Reply:
Done in 52d1e67bdb853246496504de8dd7e1978d92e119.